### ISSUE TITLE
XContentBuilder throws NumberFormatException for Date field

### ISSUE THREAD

**Issue by drmaas** [elastic/elasticsearch#15375]
This index:
```
        IndexResponse response = client.prepareIndex('twitter', 'tweet', '1')
                .setSource(jsonBuilder()
                .startObject()
                .field('user', 'dan')
                .field('postDate', new Date())
                .field('message', 'rocking out')
                .endObject()
        ).get();
```

Throws a NumberFormatException on the date field
```
MapperParsingException[failed to parse [postDate]]; nested: NumberFormatException[For input string: "2015-12-10T16:18:26.876Z"];
	at org.elasticsearch.index.mapper.FieldMapper.parse(FieldMapper.java:339)
	at org.elasticsearch.index.mapper.DocumentParser.parseObjectOrField(DocumentParser.java:314)
	at org.elasticsearch.index.mapper.DocumentParser.parseValue(DocumentParser.java:441)
	at org.elasticsearch.index.mapper.DocumentParser.parseObject(DocumentParser.java:267)
	at org.elasticsearch.index.mapper.DocumentParser.innerParseDocument(DocumentParser.java:128)
	at org.elasticsearch.index.mapper.DocumentParser.parseDocument(DocumentParser.java:79)
	at org.elasticsearch.index.mapper.DocumentMapper.parse(DocumentMapper.java:318)
	at org.elasticsearch.index.shard.IndexShard.prepareIndex(IndexShard.java:537)
	at org.elasticsearch.index.shard.IndexShard.prepareIndex(IndexShard.java:528)
	at org.elasticsearch.action.support.replication.TransportReplicationAction.prepareIndexOperationOnPrimary(TransportReplicationAction.java:1049)
	at org.elasticsearch.action.support.replication.TransportReplicationAction.executeIndexRequestOnPrimary(TransportReplicationAction.java:1060)
	at org.elasticsearch.action.index.TransportIndexAction.shardOperationOnPrimary(TransportIndexAction.java:170)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryPhase.performOnPrimary(TransportReplicationAction.java:579)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryPhase$1.doRun(TransportReplicationAction.java:452)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NumberFormatException: For input string: "2015-12-10T16:18:26.876Z"
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.lang.Long.parseLong(Long.java:589)
	at java.lang.Long.parseLong(Long.java:631)
	at org.elasticsearch.common.xcontent.support.AbstractXContentParser.longValue(AbstractXContentParser.java:145)
	at org.elasticsearch.index.mapper.core.LongFieldMapper.innerParseCreateField(LongFieldMapper.java:276)
	at org.elasticsearch.index.mapper.core.NumberFieldMapper.parseCreateField(NumberFieldMapper.java:213)
	at org.elasticsearch.index.mapper.FieldMapper.parse(FieldMapper.java:331)
	... 17 more
```

**Comment by cbuescher** [comment]
Hi @drmaas,
it looks like the `postDate` field is already mapped to `long`, so it tries to parse the date as a long value. Is it possible that you indexed a document into the same index where the `postDate` field was interpreted as a numeric value by dynamic mapping? This might also have been a document with a different type, because fields with the same name in different document types in the same index map to the same Lucene field name internally (see this [blogpost](https://www.elastic.co/blog/great-mapping-refactoring#conflicting-mappings) for more details). 
Hope this solves your problem.

**Comment by clintongormley** [comment]
As @cbuescher says, this may well be caused by having different mappings on different shards (something you can't diagnose with the get mapping API, but which is prevented in 2.0).  Closed by #8870

**Comment by drmaas** [comment]
Thanks for clarifying!
